cognite/extractorutils/rest/extractor.py

```diff
@@ -0,0 +1,160 @@
+        headers: Optional[Dict[str, Union[str, Callable[[], str]]]] = None,
+    ):
+        super(RestExtractor, self).__init__(
+            name=name,
+            description=description,
+            version=version,
+            cancelation_token=cancelation_token,
+            use_default_state_store=False,
+            config_class=RestConfig,
+        )
+        self.base_url = base_url or ""
+        self.headers = headers or {}
+        self.endpoints: List[Endpoint] = []
+
+    def endpoint(
+        self,
+        *,
+        method: HttpMethod,
+        path: str,
+        query: Dict[str, Any],
+        body: Optional[RequestBody],
+        response_type: Type[ResponseType],
+        next_url: Optional[Callable[[str, ResponseType], Optional[str]]],
+        interval: Optional[int],
+    ):
+        def decorator(func):
+            self.endpoints.append(
+                Endpoint(
+                    implementation=func,
+                    method=method,
+                    path=path,
+                    query=query,
+                    body=body,
+                    response_type=response_type,
+                    next_url=next_url,
+                    interval=interval,
+                )
+            )
+            return func
+
+        return decorator
+
+    def get(
+        self,
+        path: str,
+        *,
+        query: Optional[Dict[str, Any]] = None,
+        response_type: Type[ResponseType],
+        next_url: Optional[Callable[[HttpCall], Optional[HttpUrl]]] = None,
+        interval: Optional[int] = None,
+    ):
+        return self.endpoint(
+            method=HttpMethod.GET,
+            path=path,
+            query=query,
+            body=None,
+            response_type=response_type,
+            next_url=next_url,
+            interval=interval,
+        )
+
+    def __enter__(self) -> "RestExtractor":
+        super(RestExtractor, self).__enter__()
+        # TODO: Add more uplaod queues
+        self.event_queue = EventUploadQueue(
+            self.cognite_client, max_queue_size=10_000, max_upload_interval=60, trigger_log_level="INFO"
+        )
+        self.event_queue.__enter__()
+        return self
+
+    def __exit__(
+        self, exc_type: Optional[Type[BaseException]], exc_val: Optional[BaseException], exc_tb: Optional[TracebackType]
+    ) -> bool:
+        self.event_queue.__exit__(exc_type, exc_val, exc_tb)
+        return super(RestExtractor, self).__exit__(exc_type, exc_val, exc_tb)
+
+    def _handle_output(self, output: CdfTypes) -> None:
+        if not isinstance(output, Iterable):
+            output = [output]
+
+        peekable_output = peekable(output)
+
+        # TODO: Handle more resource types
+        if isinstance(peekable_output.peek(), Event):
+            for event in peekable_output:
+                self.event_queue.add_to_upload_queue(event)
+        else:
+            raise ValueError(f"Unexpected type: {type(output[0])}")
```

Comment:
@hakontro This is where the uploading happens. The `event_queue` will then upload whenever either enough events are added, when enough time has passed, or when the extractor shuts down.